modules/frontend/searchsharding_test.go

```diff
@@ -1022,6 +1151,151 @@ func TestSubRequestsCancelled(t *testing.T) {
 	}
 }
 
+func TestHashTraceQLQuery(t *testing.T) {
+	// exact same queries should have the same hash
+	h1 := hashForTraceQLQuery("{ span.foo = `bar` }")
+	h2 := hashForTraceQLQuery("{ span.foo = `bar` }")
+	require.Equal(t, h1, h2)
+
+	// equivalent queries should have the same hash
+	h1 = hashForTraceQLQuery("{ span.foo = `bar`     }")
+	h2 = hashForTraceQLQuery("{ span.foo = `bar` }")
+	require.Equal(t, h1, h2)
+
+	h1 = hashForTraceQLQuery("{ (span.foo = `bar`) || (span.bar = `foo`) }")
+	h2 = hashForTraceQLQuery("{ span.foo = `bar` || span.bar = `foo` }")
+	require.Equal(t, h1, h2)
+
+	// different queries should have different hashes
+	h1 = hashForTraceQLQuery("{ span.foo = `bar` }")
+	h2 = hashForTraceQLQuery("{ span.foo = `baz` }")
+	require.NotEqual(t, h1, h2)
+
+	// invalid queries should return 0
+	h1 = hashForTraceQLQuery("{ span.foo = `bar` ")
+	require.Equal(t, uint64(0), h1)
+
+	h1 = hashForTraceQLQuery("")
+	require.Equal(t, uint64(0), h1)
+}
+
+func TestCacheKeyForJob(t *testing.T) {
+	tcs := []struct {
+		name          string
+		queryHash     uint64
+		req           *tempopb.SearchRequest
+		meta          *backend.BlockMeta
+		searchPage    int
+		pagesToSearch int
+
+		expected string
+	}{
+		{
+			name:      "valid!",
+			queryHash: 42,
+			req: &tempopb.SearchRequest{
+				Start: 10,
+				End:   20,
+			},
+			meta: &backend.BlockMeta{
+				BlockID:   uuid.MustParse("00000000-0000-0000-0000-000000000123"),
+				StartTime: time.Unix(15, 0),
+				EndTime:   time.Unix(16, 0),
+			},
+			searchPage:    1,
+			pagesToSearch: 2,
+			expected:      "sj:42:00000000-0000-0000-0000-000000000123:1:2",
+		},
+		{
+			name:      "no query hash means no query cache",
+			queryHash: 0,
+			req: &tempopb.SearchRequest{
+				Start: 10,
+				End:   20,
+			},
+			meta: &backend.BlockMeta{
+				BlockID:   uuid.MustParse("00000000-0000-0000-0000-000000000123"),
+				StartTime: time.Unix(15, 0),
+				EndTime:   time.Unix(16, 0),
+			},
+			searchPage:    1,
+			pagesToSearch: 2,
+			expected:      "",
+		},
+		{
+			name:      "meta before start time",
+			queryHash: 42,
+			req: &tempopb.SearchRequest{
+				Start: 10,
+				End:   20,
+			},
+			meta: &backend.BlockMeta{
+				BlockID:   uuid.MustParse("00000000-0000-0000-0000-000000000123"),
+				StartTime: time.Unix(5, 0),
+				EndTime:   time.Unix(6, 0),
+			},
+			searchPage:    1,
+			pagesToSearch: 2,
+			expected:      "",
+		},
+		{
+			name:      "meta overlaps search start",
+			queryHash: 42,
+			req: &tempopb.SearchRequest{
+				Start: 10,
+				End:   20,
+			},
+			meta: &backend.BlockMeta{
+				BlockID:   uuid.MustParse("00000000-0000-0000-0000-000000000123"),
+				StartTime: time.Unix(5, 0),
+				EndTime:   time.Unix(15, 0),
+			},
+			searchPage:    1,
+			pagesToSearch: 2,
+			expected:      "",
+		},
+		{
+			name:      "meta overlaps search end",
```

Comment:
let's add a test case where search fits inside meta, like this:

```
start = 10
end = 20
metaStartTime = 5
metaEndTIme = 25
```

Reply:
yup, good call